prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -0,0 +1,136 @@
+import pandas as pd
+from powersimdata.utility.distance import haversine
+
+from prereise.gather.griddata.hifld import const
+from prereise.gather.griddata.hifld.data_access import load
+
+
+def floatify(value, default=float("nan")):
+    """Return a float if possible, otherwise return a default value.
+
+    :param object value: value to be coerced to float, if possible.
+    :param object default: value to be returned if float conversion isn't possible.
+    :return: (*float/object*) -- float or default value as applicable.
+    """
+    try:
+        if isinstance(value, str):
+            value = value.replace(",", "")
+        return float(value)
+    except Exception:
+        return default
+
+
+def map_generator_to_sub_by_location(generator, substation_groupby):
+    """Determine a likely substation for a generator to be connected to. Priority order
+    of mapping is: 1) if location is available and one or more substations exist in that
+    ZIP code, map by location to closest substation within that ZIP code, 2) if location
+    is available but no substations exist in that ZIP code, map to the closest
+    substation within neighboring ZIP codes, 3) if only ZIP code is available
+    (no location), and one or more substations exist, map to an arbitrarily chosen
+    substation within that ZIP code, 4) if only ZIP code is available (no location)
+    but no substations exist in that ZIP code, return NA.
+
+    :param pandas.Series generator: one generating unit from data frame.
+    :param pandas.GroupBy substation_groupby: data frame of substations, grouped by
+        (interconnect, ZIP).
+    :return: (*int/pd.NA*) -- substation ID if the generator can be mapped successfully
```

Comment:
same comment regarding NaN.

Reply:
Not sure what the best approach is when we would return an `int`. If we return `pd.NA`, we get a Series of dtype `object`, but the int values stay int (which is important when we want to use them to index into something else). If we return `float("nan")`, the Series gets converted to all floats, including the ones that were originally int:
```python
>>> import pandas as pd
>>> pd.Series([2, float("nan")])
0    2.0
1    NaN
dtype: float64
>>> pd.Series([2, pd.NA])
0       2
1    <NA>
dtype: object
```

Reply:
You are thinking in the user facing function when we map through a `lambda` function?

Reply:
Exactly:
```python
>>> import pandas as pd
>>> a = pd.Series([1,2,3])
>>> a.map(lambda x: x if x % 2 == 1 else pd.NA)
0       1
1    <NA>
2       3
dtype: object
>>> a.map(lambda x: x if x % 2 == 1 else float("nan"))
0    1.0
1    NaN
2    3.0
dtype: float64
```